kernelflinger: coreboot: minor fix on memory ranges and multiboot header

zhangckid opened the pull request:
For the details please check the patches comments.

AjanZhong commented on payloads/libpayload/arch/x86/multiboot.c:
```diff
@@ -48,6 +48,8 @@ static int mb_add_memrange(struct sysinfo_t *info, unsigned long long base,
 
 	info->memrange[info->n_memranges].base = base;
 	info->memrange[info->n_memranges].size = size;
+	if (type >= CB_MEM_VENDOR_RSVD)
```
payload defines 6 types of memory,  and loader should provide one of these 6 types. And type of memory is passed from EBX by loader, which means loader provides invalid memory type.

Why loader will provide these invalid memory type?

jsun26intel replied:
per multiboot spec:
where a value of 1 indicates available RAM, value of 3 indicates usable memory holding ACPI information, value of 4 indicates reserved memory which needs to be preserved on hibernation, value of 5 indicates a memory which is occupied by defective RAM modules and **all other values currently indicated a reserved area**.

BTW, grub defines below type:
typedef enum grub_memory_type
  {
    GRUB_MEMORY_AVAILABLE = 1,
    GRUB_MEMORY_RESERVED = 2,
    GRUB_MEMORY_ACPI = 3,
    GRUB_MEMORY_NVS = 4,
    GRUB_MEMORY_BADRAM = 5,
    GRUB_MEMORY_PERSISTENT = 7,
    GRUB_MEMORY_PERSISTENT_LEGACY = 12,
    GRUB_MEMORY_COREBOOT_TABLES = 16,
    GRUB_MEMORY_CODE = 20,
    /* This one is special: it's used internally but is never reported
       by firmware. Don't use -1 as it's used internally for other purposes. */
    GRUB_MEMORY_HOLE = -2,
    GRUB_MEMORY_MAX = 0x10000
  } grub_memory_type_t;

AjanZhong replied:
LGTM

jsun26intel replied:
@zhangckid could you please make a minor change:
          if (type >CB_MEM_VENDOR_RSVD)

AjanZhong commented on payloads/libpayload/include/sysinfo.h:
```diff
@@ -31,7 +31,7 @@
 
 /* Maximum number of memory range definitions. */
-#define SYSINFO_MAX_MEM_RANGES 32
+#define SYSINFO_MAX_MEM_RANGES 64
```
Why do we need to double this ranges? 32 ranges should enough for payload since it has been validated in both baremetal and emulator env.

jsun26intel replied:
we checked some platforms, it will exceed 32.
sometime reach 35, sometime reach 48.

find 35 memmap ranges: 
mem[0] : [ 0x0 - 0x87000], type: 0x1
mem[1] : [ 0x87000 - 0x88000], type: 0x2
mem[2] : [ 0x88000 - 0x99000], type: 0x1
mem[3] : [ 0x99000 - 0xa0000], type: 0x2
mem[4] : [ 0xa0000 - 0x100000], type: 0x2
mem[5] : [ 0x100000 - 0x100000], type: 0x1
mem[6] : [ 0x207000 - 0x400000], type: 0x1
mem[7] : [ 0x1e00000 - 0x44633000], type: 0x1
mem[8] : [ 0x44633000 - 0x44973000], type: 0x2
mem[9] : [ 0x44973000 - 0x452b8000], type: 0x2
mem[10] : [ 0x452b8000 - 0x45348000], type: 0x3
mem[11] : [ 0x45348000 - 0x45350000], type: 0x4
mem[12] : [ 0x45350000 - 0x45c00000], type: 0x2
mem[13] : [ 0x45c00000 - 0x45f0f000], type: 0x1
mem[14] : [ 0x45f0f000 - 0x45f10000], type: 0x2
mem[15] : [ 0x45f10000 - 0x45f90000], type: 0x2
mem[16] : [ 0x45f90000 - 0x45f91000], type: 0x4
mem[17] : [ 0x45f91000 - 0x45ff7000], type: 0x3
mem[18] : [ 0x45ff7000 - 0x50400000], type: 0x2
mem[19] : [ 0xfed20000 - 0xfed80000], type: 0x2
mem[20] : [ 0xff200000 - 0x100000000], type: 0x2
mem[21] : [ 0x100000000 - 0x4063f5000], type: 0x1
mem[22] : [ 0x4063f5000 - 0x40640a000], type: 0x2
mem[23] : [ 0x40640a000 - 0x4adc0a000], type: 0x2
mem[24] : [ 0x4adc0a000 - 0x4adca9000], type: 0x2
mem[25] : [ 0x4adca9000 - 0x4adcab000], type: 0x1
mem[26] : [ 0x4adcab000 - 0x4adcac000], type: 0x2
mem[27] : [ 0x4adcac000 - 0x4adcad000], type: 0x1
mem[28] : [ 0x4adcad000 - 0x4ae158000], type: 0x2
mem[29] : [ 0x4ae158000 - 0x4ae15a000], type: 0x1
mem[30] : [ 0x4ae15a000 - 0x4ae9aa000], type: 0x2
mem[31] : [ 0x4ae9aa000 - 0x4ae9ba000], type: 0x4
mem[32] : [ 0x4ae9ba000 - 0x4af804000], type: 0x2
mem[33] : [ 0x4af804000 - 0x4af80c000], type: 0x4
mem[34] : [ 0x4af80c000 - 0x4afc00000], type: 0x2

AjanZhong replied:
LGTM

AjanZhong commented on payloads/libpayload/arch/x86/head.S:
```diff
@@ -44,7 +44,7 @@ _entry:
 
 #define MB_MAGIC 0x1BADB002
-#define MB_FLAGS 0x00010003
+#define MB_FLAGS 0x00000003
```
offset 12-28 have been initialized at line 53 to line 57, they are all filled with symbols address. This multiboot header has already provided info about header/load/load end/bass end and entry address, according to multiboot spec, loader can parse these fields and take corresponding action.
If loader does not want to use these info, loader can jump to the start address of memory where payload is copied, since instruction at the start of  binary is "jump to real entry", the precondition is that payload should support PIC/PIE.

jsun26intel replied:
does this work?
It doesn't work under grub.

AjanZhong replied:
payload in coreboot woks in baremetal and qemu. also this changes breaks multiboot header definition:
magic + flag + checksum = 0.
let us discussed this patch next week.

jsun26intel replied:
line 53 to line 57 only defines:
12	u32	header_addr	if flags[16] is set
16	u32	load_addr	if flags[16] is set
20	u32	load_end_addr	if flags[16] is set
24	u32	bss_end_addr	if flags[16] is set

entry address is not specified

jsun26intel replied:
checksum is calculated by
.long -(MB_MAGIC + MB_FLAGS)
 automatically

AjanZhong replied:
Based on discussion this afternoon, I think root cause of zeroed header address should be caused by: 
CONFIG_LP_BASE_ADDRESS is unset.
According to link script payloads/libpayload/arch/x86/libpayload.ldscript, please have a try on set CONFIG_LP_BASE_ADDRESS  with specified address.

jsun26intel replied:
you are right. I tried it works.

Let's hold on this patch.

AjanZhong reviewed (commented):
Please remove abundant space/blank if comment message of "kernelflinger: coreboot: minor fix on memory ranges"

Aryan12138 committed:
coreboot: minor fix on memory ranges

1. enlarge max memory range entries to 64
2. For the unknown memory type which bootloader provides,
set it to CB_MEM_RESERVED

Tracked-On: OAM-111353

Signed-off-by: Victor Sun <victor.sun@intel.com>
Signed-off-by: Zhang Chen <chen.zhang@intel.com>

zhangckid commented:
> 

Done.

zhangckid commented:
> Please remove abundant space/blank if comment message of "kernelflinger: coreboot: minor fix on memory ranges"

Address this comments and update new version.

AjanZhong reviewed (approved):
LGTM
